observer/monitor.go

```diff
@@ -0,0 +1,39 @@
+package observer
+
+import (
+	"strconv"
+	"time"
+
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/observer/probers"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type monitor struct {
+	period time.Duration
+	prober probers.Prober
+	logger blog.Logger
+	metric prometheus.Registerer
+}
+
+// start spins off a goroutine that attempts to probe for each period
+// specified in the monitor config.
+func (m monitor) start() {
+	go func() {
+		for {
+			select {
+			case <-time.NewTicker(m.period).C:
+				// Attempt to probe the configured target.
+				success, dur := m.prober.Probe(m.period)
```

Comment:
Here we use the configured period as the timeout for the probe attempt. Assume that every probe attempt is timing out (taking the full e.g. 5 seconds). Then this will slowly get further and further behind: the ticker goroutine will try to write a tick to the channel every 5 seconds, but the timeout plus additional processing around it will be taking 5.1 seconds. I don't think this will cause anything to *fail* (the ticker should happily block on trying to send to the channel if its previous tick hasn't yet been received), but it may cause the metrics to be confusing to analyze. Maybe the default timeout should be `m.period / 2` or something like that?

Reply:
In this scenario the data-point exceeds the partition of the period for which it was called, which causes jitter equivalent to the time to schedule the go-routine.  If it continues to fail, this is fine, the problem I see with this behavior is flapping. Any time that a failed data-point is followed by a succeeding data-point the resolution for that period is doubled. However, we don't typically want to alarm on a single data-point. We typically alarm on a percentage of failed data-points over a fixed period or a number of failing contiguous data-points. Additionally, most of these metrics will never be used (or collected??) at sub-second resolution. Still, your point is correct.

We have a couple options:
1. Accept this behavior to save ourselves from adding complexity (but note it)
2. Fix this behavior by using contexts and deadlines

I don't really have a strong opinion either way. When Jacob and I talked about it in an earlier review we opted for reduced complexity, but our discussion was not specific to the observability issue you've highlighted here.

Reply:
Actually this question (timeout vs period) raises an interesting issue: I think we might like to do some of these probes as often as (say) once per second. But if we want a timeout of 10s, which is what we currently use for some of our blackbox_exporter metrics, that would conflict with the current implementation. There's a way to fix that - run each probe on a goroutine. But I don't think that's needed for this first iteration. I would go with the hardcoded `timeout == period` for now, with the expectation that in the future we may allow timeout to be configured separately from period.

Reply:
Alright, @aarongable and I talked about this 1:1 and we realized that while I specify a probe interval of `period` I am mistakenly treating it as a timeout. Now the idea of deriving the timeout from the period halved totally makes sense :)